pkg/reconciler/finalizer/finalizer.go

```diff
@@ -0,0 +1,298 @@
+		return errors.Wrap(err, "removing Finalizing condition")
+	}
+	if _, err := f.RemoveFinalizer(ctx, obj); err != nil {
+		return errors.Wrap(err, "removing finalizer")
+	}
+	return nil
+}
+
+// AddFinalizer adds the `configsync.gke.io/reconciler` finalizer to the specified
+// object.
+func (f *Finalizer) AddFinalizer(ctx context.Context, obj *unstructured.Unstructured) (bool, error) {
+	updated, err := updateObjectWithRetry(ctx, f.Client, obj, func() error {
+		if !controllerutil.AddFinalizer(obj, metadata.ReconcilerFinalizer) {
+			// Already added. No change necessary.
+			return &NoUpdateError{}
+		}
+		return nil
+	})
+	if err != nil {
+		return updated, errors.Wrapf(err, "failed to remove finalizer")
+	}
+	if updated {
+		klog.Info("Finalizer injection successful")
+	} else {
+		klog.Warning("Finalizer injection skipped: already exists")
+	}
+	return updated, nil
+}
+
+// RemoveFinalizer removes the `configsync.gke.io/reconciler` finalizer from the
+// specified object.
+func (f *Finalizer) RemoveFinalizer(ctx context.Context, obj *unstructured.Unstructured) (bool, error) {
+	updated, err := updateObjectWithRetry(ctx, f.Client, obj, func() error {
+		if !controllerutil.RemoveFinalizer(obj, metadata.ReconcilerFinalizer) {
+			// Already removed. No change necessary.
+			return &NoUpdateError{}
+		}
+		return nil
+	})
+	if err != nil {
+		return updated, errors.Wrapf(err, "failed to remove finalizer")
+	}
+	if updated {
+		klog.Info("Finalizer removal successful")
+	} else {
+		klog.Warning("Finalizer removal skipped: already removed")
+	}
+	return updated, nil
+}
+
+// setFinalizingCondition sets the ReconcilerFinalizing condition on the
+// specified object.
+func (f *Finalizer) setFinalizingCondition(ctx context.Context, obj *unstructured.Unstructured) (bool, error) {
+	updated, err := updateObjectWithRetry(ctx, f.Client, obj, func() error {
+		// Convert to RootSync/RepoSync to make it easier to update the conditions.
+		// TODO: Optimize by avoiding multiple type conversions.
+		rObj, err := kinds.ToTypedObject(obj, core.Scheme)
+		if err != nil {
+			return err
+		}
+		switch tObj := rObj.(type) {
+		case *v1beta1.RootSync:
+			if !rootsync.SetReconcilerFinalizing(tObj, "ResourcesDeleting", "Deleting managed resource objects") {
+				// Already removed. No change necessary.
+				return &NoUpdateError{}
+			}
+		case *v1beta1.RepoSync:
+			if !reposync.SetReconcilerFinalizing(tObj, "ResourcesDeleting", "Deleting managed resource objects") {
+				// Already removed. No change necessary.
+				return &NoUpdateError{}
+			}
+		case client.Object:
+			return fmt.Errorf("invalid object type: expected v1beta1.RootSync or v1beta1.RepoSync: %s", objSummary(tObj))
+		default:
+			return fmt.Errorf("invalid object type: expected v1beta1.RootSync or v1beta1.RepoSync: %T", rObj)
+		}
+		// Copy updates back into the original unstructured object
+		uObj, err := kinds.ToUnstructured(obj, core.Scheme)
```

Comment:
This should be `rObj`.

Reply:
Good catch. I obviously need some tests to validate this condition.